src/FSharp.Core/list.fs

```diff
@@ -987,3 +988,174 @@ module List =
+
+        let i = random.Next(0, inputLength)
+        source[i]
+
+    [<CompiledName("RandomChoiceBy")>]
+    let randomChoiceBy (randomizer: unit -> float) (source: 'T list) : 'T =
+        let inputLength = source.Length
+
+        if inputLength = 0 then
+            invalidArg "source" LanguagePrimitives.ErrorStrings.InputSequenceEmptyString
+
+        let i = Microsoft.FSharp.Primitives.Basics.Random.next randomizer 0 inputLength
+        source[i]
+
+    [<CompiledName("RandomChoice")>]
+    let randomChoice (source: 'T list) : 'T =
+        randomChoiceWith ThreadSafeRandom.Shared source
+
+    [<CompiledName("RandomChoicesWith")>]
+    let randomChoicesWith (random: Random) (count: int) (source: 'T list) : 'T list =
+        checkNonNull "random" random
+
+        if count < 0 then
+            invalidArgInputMustBeNonNegative "count" count
+
+        let inputLength = source.Length
+
+        if inputLength = 0 then
+            invalidArg "source" LanguagePrimitives.ErrorStrings.InputSequenceEmptyString
+
+        [
+            for _ = 0 to count - 1 do
+                let j = random.Next(0, inputLength)
+                source[j]
+        ]
+
+    [<CompiledName("RandomChoicesBy")>]
+    let randomChoicesBy (randomizer: unit -> float) (count: int) (source: 'T list) : 'T list =
+        if count < 0 then
+            invalidArgInputMustBeNonNegative "count" count
+
+        let inputLength = source.Length
+
+        if inputLength = 0 then
+            invalidArg "source" LanguagePrimitives.ErrorStrings.InputSequenceEmptyString
+
+        [
+            for _ = 0 to count - 1 do
+                let j = Microsoft.FSharp.Primitives.Basics.Random.next randomizer 0 inputLength
+                source[j]
+        ]
+
+    [<CompiledName("RandomChoices")>]
+    let randomChoices (count: int) (source: 'T list) : 'T list =
+        randomChoicesWith ThreadSafeRandom.Shared count source
+
+    [<CompiledName("RandomSampleWith")>]
+    let randomSampleWith (random: Random) (count: int) (source: 'T list) : 'T list =
+        checkNonNull "random" random
+
+        if count < 0 then
+            invalidArgInputMustBeNonNegative "count" count
+
+        let inputLength = source.Length
+
+        if inputLength = 0 then
+            invalidArg "source" LanguagePrimitives.ErrorStrings.InputSequenceEmptyString
+
+        if count >= inputLength then
+            invalidArgOutOfRange "count" count "source.Length" inputLength
+
+        // algorithm taken from https://github.com/python/cpython/blob/main/Lib/random.py
+        let setSize =
+            Microsoft.FSharp.Primitives.Basics.Random.getMaxSetSizeForSampling count
+
+        if inputLength <= setSize then
+            let pool = source |> toArray
+
+            [
+                for i = 0 to count - 1 do
+                    let j = random.Next(0, inputLength - i)
+                    yield pool[j]
+                    pool[j] <- pool[inputLength - i - 1]
+            ]
```

Comment:
For #16948 and #17067 to kick in (once I fix the aforementioned hole where they don't apply to implicit yields), I think this should probably do something like:

```suggestion
            [
                for i = 0 to count - 1 do
                    let j = random.Next(0, inputLength - i)
                    let x = pool[j]
                    pool[j] <- pool[inputLength - i - 1]
                    yield x
            ]
```

I could in theory also make that optimization kick in even with trailing sequential expressions, but that might make less sense.

Reply:
Fixed